onlinejudge/service/toph.py

```diff
@@ -0,0 +1,172 @@
+# Python Version: 3.x
+import posixpath
+import re
+import string
+import urllib.parse
+from typing import *
+
+import bs4
+import requests
+
+import onlinejudge.dispatch
+import onlinejudge.implementation.logging as log
+import onlinejudge.implementation.utils as utils
+import onlinejudge.type
+from onlinejudge.type import SubmissionError
+
+
+@utils.singleton
+class TophService(onlinejudge.type.Service):
+    def login(self, get_credentials: onlinejudge.type.CredentialsProvider, session: Optional[requests.Session] = None) -> bool:
+        session = session or utils.new_default_session()
+        url = 'https://toph.co/login'
+        # get
+        resp = utils.request('GET', url, session=session)
+        if resp.url != url:  # redirected
+            log.info('You are already logged in.')
+            return True
+        # parse
+        soup = bs4.BeautifulSoup(resp.content.decode(resp.encoding), utils.html_parser)
+        form = soup.find('form', class_='login-form')
+        log.debug('form: %s', str(form))
+        username, password = get_credentials()
+        form['action'] = '/login' # to avoid KeyError inside form.request method as Toph does not have any defined action
+        form = utils.FormSender(form, url=resp.url)
+        form.set('handle', username)
+        form.set('password', password)
+        # post
+        resp = form.request(session)
+        resp.raise_for_status()
+
+        resp = utils.request('GET', url, session=session)    # Toph's Location header is not getting the expected value
+        if resp.url != url:
+            log.success('Welcome, %s.', username)
+            return True
+        else:
+            log.failure('Invalid handle/email or password.')
+            return False
+
+    def is_logged_in(self, session: Optional[requests.Session] = None) -> bool:
+        session = session or utils.new_default_session()
+        url = 'https://toph.co/login'
+        resp = utils.request('GET', url, session=session, allow_redirects=False)
+        return resp.status_code != 200
+
+    def get_url(self) -> str:
+        return 'https://toph.co/'
+
+    def get_name(self) -> str:
+        return 'toph'
+
+    @classmethod
+    def from_url(cls, s: str) -> Optional['TophService']:
+        # example: https://toph.co/
+        # example: http://toph.co/
+        result = urllib.parse.urlparse(s)
+        if result.scheme in ('', 'http', 'https') \
+                and result.netloc == 'toph.co':
+            return cls()
+        return None
+
+class TophProblem(onlinejudge.type.Problem):
+    def __init__(self, problem_id: str, kind: Optional[str] = None, contest_id: Optional[str] = None):
```

Comment:
The variable `kind` seems to be not necessary for now (`contest_id` is `None` implies it is `problem`, and not `None` does `contest`). 

I think that adding this variable is unnecessary generalization, and removing this is simple and better, but we should add this if there is another existing mode or a plan to add a new mode.
@kfaRabi Do you know another mode or a plan?

Reply:
@kmyk Yeah you are right. This is really unnecessary. I kept it so that I could avoid writing any if/else condition :stuck_out_tongue_closed_eyes:  I have removed this and added two more submission tests using CPP (Kept the old python ones). Please let me know if there is anything else that I could improve. I would be happy to do so :slightly_smiling_face:

Reply:
You are right. We do not need this as Toph does not plan to add any other mode soon.